picasso/src/main/java/com/squareup/picasso/SourceMarker.java

```diff
@@ -0,0 +1,185 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.picasso;
+
+import java.io.IOException;
+import okio.Buffer;
+import okio.BufferedSource;
+import okio.ForwardingSource;
+import okio.Okio;
+import okio.Source;
+
+/**
+ * Builds a buffered source that can rewind to a marked position earlier in the stream.
+ *
+ * <p>Mark potential positions to rewind back to with {@link #mark}; rewind back to these positions
+ * with {@link #reset}. Both operations apply to the position in the {@linkplain #source() buffered
+ * source}; resetting will impact the buffer.
+ *
+ * <p>When marking it is necessary to specify how much data to retain. Once you advance above this
+ * limit, the mark is discarded and resetting is not permitted. This may be used to lookahead a
+ * fixed number of bytes without loading an entire stream into memory. To reset an arbitrary
+ * number of bytes use {@code mark(Long#MAX_VALUE)}.
+ */
+public final class SourceMarker {
+
+  /*
+   * This class wraps the underlying source in a MarkSource to support mark and reset. It creates a
+   * BufferedSource for the caller so that it can track its offsets and manipulate its buffer.
+   */
+
+  /**
+   * The offset into the underlying source. To compute the user's offset start with this and
+   * subtract userBuffer.size().
+   */
+  long offset;
+
+  /** The offset of the earliest mark, or -1 for no mark. */
+  long mark = -1L;
+
+  /** The offset of the latest readLimit, or -1 for no mark. */
+  long limit = -1L;
+
+  boolean closed;
+
+  final MarkSource markSource;
+  final BufferedSource userSource;
+
+  /** A copy of the underlying source's data beginning at {@code mark}. */
+  final Buffer markBuffer;
+
+  /** Just the userSource's buffer. */
+  final Buffer userBuffer;
+
+  public SourceMarker(Source source) {
+    this.markSource = new MarkSource(source);
+    this.markBuffer = new Buffer();
+    this.userSource = Okio.buffer(markSource);
+    this.userBuffer = userSource.buffer();
+  }
+
+  public BufferedSource source() {
+    return userSource;
+  }
+
+  /**
+   * Marks the current position in the stream as one to potentially return back to. Returns the
+   * offset of this position. Call {@link #reset(long)} with this position to return to it later. It
+   * is an error to call {@link #reset(long)} after consuming more than {@code readLimit} bytes from
+   * {@linkplain #source() the source}.
+   */
+  public long mark(long readLimit) throws IOException {
```

Comment:
I don't think `readLimit` is useful here. We don't know how much the image decoder will consume in order to discern the size of the image. Picasso will always pass `Long.MAX_VALUE` to this to allow reading unlimited bytes up to EOF.

Reply:
Agreed for Picasso. I think I wanna promote this into Okio samples, or perhaps elsewhere.